src/geometry/manifold/cont_mdiff.lean

```diff
@@ -1640,3 +1648,161 @@ hf.smul hg
 lemma smooth.smul {f : M → 𝕜} {g : M → V} (hf : smooth I 𝓘(𝕜) f) (hg : smooth I 𝓘(𝕜, V) g) :
   smooth I 𝓘(𝕜, V) (λ p, f p • g p) :=
 hf.smul hg
+
+/-! ### Smoothness of (local) structomorphisms -/
+section
+variables [charted_space H M'] [IsM' : smooth_manifold_with_corners I M']
+include Is IsM'
+
+/-- Let `M` and `M'` be smooth manifolds with the same model-with-corners, `I`.  Then `f : M → M'`
+is a local structomorphism for `I`, if and only if it is manifold-smooth on the domain of definition
+in both directions. -/
+lemma is_local_structomorph_on_cont_diff_groupoid_iff (f : local_homeomorph M M') :
+  lift_prop_on (cont_diff_groupoid ⊤ I).is_local_structomorph_within_at f f.source
+  ↔ smooth_on I I f f.source ∧ smooth_on I I f.symm f.target :=
+begin
+  split,
+  { -- We first show that a local homeomorph `f : M → M'` which is a local structomorphism is smooth
+    -- in both directions.
+    intros h,
+    split,
```

Comment:
It gives the impression that you will do twice exactly the same thing. What about proving first that if `f` is a local_structomorph on `f.source`, then `f.symm` is a local_structomorph on `f.target`, and then using a common lemma for both directions?

Reply:
Oops, I have just realized that Floris has already made the same comment a few days ago. It probably means that this is a good idea :-)